### ISSUE TITLE
Unable to use certain Other Quote Characters

### ISSUE THREAD

**Issue by peachtart** [sqlitebrowser/sqlitebrowser#1860]
Details for the issue
--------------------

I'm trying to use some non-standard delimiters to import a 'CSV', but I'm having some issues with using the Other quote delimiter option.

Specifically I am trying to use the characters:

- ` aka ASCII 20`
- `þ aka ASCII 254`

as the comma and quote respectively. The comma seems fine, but the quote is causing me issues. Using regular double quotes as the Other quote character on an appropriately edited file seems to work fine.

![Screen Shot 2019-04-26 at 2 25 53 PM](https://user-images.githubusercontent.com/26336313/56829296-588f8e00-6831-11e9-96cd-20d3c33f5e3f.png)

As per the screenshot, the þ is not being cleaned out in the preview and also remains in the fields post import. I did a quick test using the (arbitrarily chosen) `é` character as a delimiter and ran into similar issues.

Thanks!

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [x] Windows: ( _version:_ ___ )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- [x] 3.11.2
- [ ] 3.11.1
- [ ] 3.10.1
- [ ] Other: ___

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by mgrojo** [comment]
ASCII has only 7 bits and consequently cannot have a code for 254. You're referring to ISO-8859-1 (aka Latin-1), where code 254 is actually þ. Try with that encoding and see what happens. If still doesn't work, be sure to use the appropiate encoding for the file, but if þ only gets a byte, that's not UTF-8.

**Comment by peachtart** [comment]
Ah, so the 7 bit ASCII characters are supported as delimiters?

No luck with an ISO-8859-1 file either. I had been using a UTF-8 file previously and typing in the delimiter. I attached some example files for reference.

[Example.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/3122839/Example.zip)

**Comment by scottfurry** [comment]
sqlitebrowser doesn't use UTF-8? Or UTF-16? We're still hamstrung by 1960's character encoding?

Sqlite C-API can handle it (see: https://www.sqlite.org/c3ref/open.html). Windows can handle UTF-16(see: https://en.wikipedia.org/wiki/Unicode_in_Microsoft_Windows). I'm sure we can wrangle bits better to bump up the encoding. 

Is there some aspect to Qt library missed? (Doubtful).

Can we use this bug(maybe reworded a bit) with a target of next major(4.0)/minor release(3.12/13) release.
Recommended title edit - Ensure proper character handling on CSV imports.
Code edit/testing to accommodate:
- ISO-8859-1 (Windows)
- UTF-8(Linux)
- UTF-16(MAC/Windows)

**Comment by mgrojo** [comment]
sqlitebrowser has good support for encoding. It's just a problem with the quoting character. Characters outside the US-ASCII (7 bits) are not currently supported as quoting character in the CSV import. The file has to be edited before passing it to the application. 

There shouldn't be too difficult to support at least Latin-1, but it should be done without breaking the optimizations made by @MKleusberg in the CSV importing code. Nevertheless, 3.10.1 was also affected.

**Comment by MKleusberg** [comment]
@peachtart I think it should work with tomorrow's nightly build. You can download it [here](https://nightlies.sqlitebrowser.org/latest/). Are you interested in trying that out? :smile: 

I hope I didn't break anything else in the process. So any tests with any CSV files are welcomed :wink: I also did a quick benchmark using the postcodes files (636MB) and the time went up from 21s to 22s. So it does cost us a bit of time but I think it's negligible for now.

**Comment by peachtart** [comment]
No luck with the 5/9 build using the characters mentioned above. Additionally I noticed that if I reopen the import dialog, it looks like the character used previously is being misremembered. Any chance something is getting lost in translation by this text box?

Specifically:
1. Attempt to import.
![Screen Shot 2019-05-09 at 10 22 50 AM](https://user-images.githubusercontent.com/26336313/57481992-787f7280-7271-11e9-8030-141b5fbcbc00.png)
2. Cancel the dialog.
3. Reopen the import dialog.
![Screen Shot 2019-05-09 at 10 23 49 AM](https://user-images.githubusercontent.com/26336313/57481997-7e755380-7271-11e9-81e2-b9b60bc5a7cd.png)

**Comment by MKleusberg** [comment]
You're right. For quote characters the saving & restoring of the "Other" character didn't work. I did all my testing with the separator character :wink: That should be fixed in tomorrow's nightly build.

As for the actual import: Do you still have problems with using the ```þ``` character for quotes? This should definitely work now. I used it for testing all the time. But you definitely need to select the right encoding (likely ISO-8859-1).

The new quote character you show in your last screenshots won't work though. We are using UTF8 internally and before the changes we only supported characters which take up to 7 bits in UTF8 (which is effectively ASCII). Now we support up to 16 bits in UTF8. But ``ʒ+`̌``` is a combining character and requires 32 bits. Do you need support for that character as well or was it just for testing? :smile:

**Comment by peachtart** [comment]
Yep, that character was chosen arbitrarily. I had only tested UTF8 and UTF16 before with the original characters which is when I still saw the issue with the `þ`. I just tested again with the 5/14 build and it looks like UTF8 and UTF16 are working as well.

It is however still remembering the wrong character after reopening--I've been using the .zip versions from the latest build, but is there some setting I should clear out to try to see if that is fixed?

**Comment by mgrojo** [comment]
> It is however still remembering the wrong character after reopening--I've been using the .zip versions from the latest build, but is there some setting I should clear out to try to see if that is fixed?

I don't know whether it would make any change, but try with "Edit > Preferences > Restore Defaults".

**Comment by peachtart** [comment]
That worked. Thanks!

My scenario seems fully accounted for in the latest build I downloaded. I'll let you all decide how to handle closing out this issue.

Thanks again.

**Comment by mgrojo** [comment]
Thanks, @peachtart. I think this can be closed now.
